### ISSUE TITLE
Unable to create native Threads in OCP containers

### ISSUE THREAD

**Issue by yasassri** [openshift/origin#24360]
In OCP 4.2 we are unable to create many native threads.

##### Version
4.2

### Steps To Reproduce
1. Create a container with JDK and execute the following java class.
```
public class TLimitCheck{

  public static void main(String[] args) {
    int count = 0;
    while (true) {
      count++;     
      new Thread(new Runnable() {
        public void run() {
          try {
            Thread.sleep(50000);
          } catch (Exception e) {}
        }
      }).start();
      System.out.println("Current Thread: " + count);
    }
  }
}
```
2. The container will hit the process limit after creating arond 1000 Threads.

### Description

Recently we were getting the following exception in one of our Containers which was running a Java application. This container used to run perfectly on Openshift 3.11.

```
Exception in thread "main" java.lang.OutOfMemoryError: unable to create new native thread
        at java.lang.Thread.start0(Native Method)
        at java.lang.Thread.start(Thread.java:717
```

As seen below, when you execute the aforementioned code in Openshift 3.11, it used to create 4000+ Threads. But in Openshift 4.2 it just created around 1000 Threads.

OCP 3.11
<img width="1197" alt="OS_311" src="https://user-images.githubusercontent.com/7681361/71708107-da9a7280-2e14-11ea-8a6d-37f8658ce08b.png">

OCP 4.2
<img width="1021" alt="OS_4 2" src="https://user-images.githubusercontent.com/7681361/71708121-ea19bb80-2e14-11ea-8232-79e1cd5ce5fc.png">

Also following are the ulimits in the container OS.
![limits](https://user-images.githubusercontent.com/7681361/71708169-3b29af80-2e15-11ea-8028-27995d05bef4.png)

Note: I used Openshift Playground for the above Tests.

### Questions
1. What has changed in OCP 4.2 and how can I increase this limit?

**Comment by benhwebster** [comment]
Anyone still puzzled by this, see this RH KB: https://access.redhat.com/solutions/5366631

TLDR or for those without access: You need to update both the kubelet config and crio config to change max PID limits (default of 1024 in 4.x) 

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: worker-kubeconfig-fix
spec:
  machineConfigPoolSelector:
    matchLabels:
      custom-kubelet: small-pods 
  kubeletConfig:
      podPidsLimit: 4096 
```

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: ContainerRuntimeConfig
metadata:
 name: set-pids-limit
spec:
 machineConfigPoolSelector:
   matchLabels:
     custom-crio: high-pid-limit
 containerRuntimeConfig:
   pidsLimit: -1 
```
Changing the crio config to -1 allows unlimited threads, or in this example 4096 as limited by the kubelet config above.
